text/0193-removal-of-type-unions.md

```diff
@@ -0,0 +1,575 @@
+  return by the getter
+- **Java** generics are _erased_ during compilation (where they are _reified_ in **C#**): overloads that differ only in
+  generic arguments (e.g: `void call(List<String>)` and `void call(List<Integer>)`) are invalid and will cause a
+  compilation failure, since they share the same _erasure_ (e.g: `void call(List)`) and a _Java Virtual Machine_ would
+  be unable to determine which implementation to dynamically dispatch to.
+
+Additionally, certain static languages (such as **Go**) do not support overloading at all, rendering this approach
+entirely unsuitable.
+
+# Design Summary
+
+Removal of _type unions_ from the AWS CDK requires designing new APIs for those places where they are in use. This
+includes two kinds of APIs:
+
+- [Generated as part of the CloudFormation Resources](#generated-code)
+- [Manually authored as part of the AWS Construct Library or AWS CDK Core framework](#hand-written-code)
+
+# Detailed Design
+
+## Generated code
+
+As previously discussed, the _CloudFormation Resources_ ubiquitously rely on _type unions_ to allow providing late-bound
+values to the various properties accepted by the resources. Late-binding values is necessary to support mutation APIs
+which provide enhanced ergonomics for certain constructs, particularly where various components can be expected to
+collaborate on the finalized configuration of a resource.
+
+### Where existing APIs already support the use-case in a safe manner
+
+All _type unions_ where late-bound values can be represented in the literal type of the value (`string`, `number` and
+`any`) can be simplified to that literal type, since the required APIs already exist and the migration path for those
+use-cases is already very well defined: using `Lazy` factory methods to manually wrap `IResolvable` instances:
+- Current instances of:
+  ```ts
+  public propertyName: string | IResolvable;
+  ```
+
+- Are changed to:
+  ```ts
+  public propertyName: string;
+  ```
+
+- And places where an `IResolvable` was passed (if any) are replaced to:
+  ```ts
+  let value: IResolvable;
+  // ...
+  { propertyName: Lazy.stringValue(value) }
+  ```
+
+### Use-cases where existing APIs are not usable from statically typed languages
+
+On the other hand, _type unions_ of `IResolvable` and complex types (including collections) cannot be resolved that
+easily. In **TypeScript** and other dynamic languages (such as **Python**), the `Lazy.asAny` method can be used to wrap
+an `IResolvable` and evade the type checker. This approach is however no workable in **C#**, **Go** and **Java** (and
+likely other statically typed languages we may wish to support in the future). For example the `s3.CfnBucket` accepts
+(and exposes) properties such as:
+```ts
+lifecycleConfiguration: CfnBucket.LifecycleConfigurationProperty | cdk.IResolvable | undefined;
+```
+
+#### Recommended Approach
+
```

Comment:
I would lead with the more common TypeScript use case (L2s) and show how this can be resolved there.